File: mm485/arduino/packet.cpp
/*
 * packet.cpp
 *
 *  Created on: May 15, 2016
 *      Author: sebastiano
 */

#include "packet.h"
#include "string.h"
#include "stdio.h"

Packet::Packet() {
	clear();
}

Packet::Packet(uint8_t src, uint8_t dst, const unsigned char* d, size_t size) {
	source = src;
	dest = dst;
	length = size;
	memcpy(data, d, length);
	packet_id = id_calculate();
	crc = crc_calculate();
}

Packet::Packet(uint8_t src, uint8_t dst, uint16_t id, const unsigned char* d, size_t size) {
	source = src;
	dest = dst;
	length = size;
	memcpy(data, d, length);
	packet_id = id;
	crc = crc_calculate();
}

uint16_t Packet::id_calculate() {
	return crc_calculate();
}

uint16_t Packet::crc_calculate() {
	size_t s_len = 2+length;
	char s[s_len];

	s[0] = dest;
	s[1] = length;
	memcpy((uint8_t *)(s+2), data, length);

    return ModRTU_CRC(s, s_len);
}

bool Packet::validate() {
	return crc == crc_calculate();
}

void Packet::clear() {
	source = dest = 0;source = dest = packet_id = length = data[0] = crc = retry = timeout = 0;
}

bool Packet::is_empty() {
	return source == 0 && dest == 0;
}

Packet* Packet::deserialize(const unsigned char* msg) {
	source = msg[0];
	dest = msg[1];
	packet_id = ((unsigned char)msg[2] << 8) | (unsigned char)msg[3];
	length = msg[4];
	memcpy(data, (char *)(msg+5), length);
	crc = ((unsigned char)msg[5+length] << 8) | (unsigned char)msg[6 + length];

	return this;
}

size_t Packet::serialize(unsigned char *msg) {
	msg[0] = source;
	msg[1] = dest;
	msg[2] = (unsigned char)((packet_id >> 8) & 0xff);
	msg[3] = (unsigned char)(packet_id & 0xff);
	msg[4] = length;
	memcpy((char *)(msg+5), data, length);
	msg[5+length] = (unsigned char)((crc >> 8) & 0xff);
	msg[6+length] = (unsigned char)(crc & 0xff);
	msg[7+length] = EOP;

	return 8+length;
}

File: mm485/packet.py
import logging

from PyCRC.CRC16 import CRC16


def enc128(data):
    def enc(_c, _n, _msb):
        _lsb = ((_c << _n) | _msb) & 127
        _msb = _c >> (7 - _n)
        return _lsb, _msb

    v = []
    n = 0
    msb = 0
    for c in data:
        lsb, msb = enc(c, n, msb)
        v.append(lsb)
        if n < 7:
            n += 1
        else:
            lsb, msb = enc(msb, 0, 0)
            v.append(lsb)
            n = 1
    # if msb or len(v) < round(((len(data) / 7) * 8) + 0.5):
    if msb:
        v.append(msb)
    return v


def dec128(data):
    v = []
    n = 1
    lsb = data[0]
    for c in data[1:]:
        msb = ((c << (8 - n)) | lsb) & 255
        lsb = c >> n
        if n != 0:
            v.append(msb)
        n = n + 1 if n < 7 else 0
    if lsb:
        v.append(lsb)
    return v


class Packet(object):
    ACK = b'\x7d'
    ERR = b'\x7e'
    EOP = b'\xfe'  # End of packet
    EOM = b'\xff'  # End Of message

    source = ''
    dest = ''
    packet_id = ''
    data = ''
    length = ''
    crc = ''

    def __init__(self, source=None, dest=None, data=None, packet_id=None, length=None, crc=None):
        self.retry = 0
        self.timeout = 0
        self.logextra = {'node': source}
        if source is not None and dest is not None and data is not None:
            self.source = source if isinstance(source, bytes) else bytes([source])
            self.dest = dest if isinstance(dest, bytes) else bytes([dest])
            self.data = bytearray(data, "utf-8") if isinstance(data, str) else data
            self.length = bytes([len(data)]) if length is None else bytes([length])
            self.packet_id = self.id_calculate() if packet_id is None else (
                packet_id if isinstance(packet_id, bytes) else bytes([packet_id])
            )
            self.crc = self.crc_calculate() if crc is None else crc

    def __str__(self):
        return str(self.__dict__)

    def __eq__(self, other):
        return self.serialize() == other.serialize()

    def id_calculate(self):
        return self.crc_calculate()

    def crc_calculate(self):
        crc = 0
        try:
            crc = CRC16(modbus_flag=True).calculate(self.dest + self.length + bytes(self.data))
            crc = crc.to_bytes(2, byteorder='little')
        except Exception as e:
            logging.error("CRC error: %s", e, extra=self.logextra)
        return crc

    def validate(self):
        return self.crc == self.crc_calculate()

    def deserialize(self, msg):
        try:
            self.crc = bytes([msg[0], msg[1]])
            self.packet_id = bytes([msg[2], msg[3]])
            self.source = bytes([msg[4]])
            self.dest = bytes([msg[5]])
            self.length = bytes([msg[6]])
            self.data = msg[7:7 + self.length[0]]
            # self.data = bytes(dec128(msg[5:5 + msg[4]]))
            # self.length = bytes([len(self.data)])
            logging.debug("Deserialized packet: %s", self.__str__(), extra=self.logextra)
        except Exception as e:
            logging.error("Error deserialization of %s : %s", msg, e, extra=self.logextra)
        return self

    def serialize(self):
        return self.crc + self.packet_id + self.source + self.dest + self.length + self.data + self.EOP

    def decode(self, data):
        logging.info('Decode stream %s', data, extra=self.logextra)
        decoded = bytes(dec128(data))
        logging.debug("Decoded stream: %s", decoded, extra=self.logextra)
        return self.deserialize(decoded)

    def encode(self):
        serialized = self.serialize()
        logging.info('Encode stream %s', serialized, extra=self.logextra)
        encoded = bytes(enc128(serialized))
        logging.debug("Encoded stream: %s", encoded, extra=self.logextra)
        return encoded